sling/nlp/wikicat/browser.py

```diff
@@ -0,0 +1,889 @@
+      ["Signature", "Example Category", "Score", "#Members / #Categories"],
+      [t.name for t in FactMatchType])
+
+    for signature, stats in all_stats:
+      self._begin("tr")
+      self._begin("td")
+      self._form_anchor(signature, signature)
+      self._end("td")
+      self._begin("td")
+      self._form_anchor(stats.example_category.name, stats.example_qid)
+      self._end("td")
+      self._cell(stats.score, numeric=True)
+      self._cell("%d / %d" % (stats.members, stats.num), numeric=True)
+      self._separator(header=False)
+      self.write_fact_match_counts(stats.fact_stats)
+      self._end("tr")
+    self._end("table")
+
+
+  # Handler for signature inputs.
+  def handle_signature(self, signature, categories, form):
+    score_type = form.getvalue("main_form_sort_metric")
+    fact_weights = self.fact_match_weights(form)
+    signature_type = form.getvalue("main_form_signature_type")
+
+    # Sort all parses with this signature.
+    output = []
+    for (qid, category, parse) in categories:
+      score = self.parse_score(category, parse, score_type, fact_weights)
+      output.append((qid, category, parse, score))
+    output.sort(key=lambda x: -x[3])
+
+    # Get fact-matching statistics. Consider only the top parse for a category
+    # if it has >1 parses with the same signature.
+    category_count = defaultdict(int)
+    match_counts = defaultdict(int)
+    span_match_counts = defaultdict(lambda: defaultdict(int))
+    num_members = 0
+    for qid, category, parse, score in output:
+      category_count[qid] += 1
+      if category_count[qid] > 1:
+        continue
+      num_members += len(category.members)
+      util.fact_matches_for_parse(parse, match_counts)
+      for span in parse.spans:
+        span_signature = util.span_signature(span, signature_type)
+        util.fact_matches_for_span(span, span_match_counts[span_signature])
+
+    # Write a summary.
+    self._tag("div",
+      "<b>%s</b>: in %d (category, parse) pairs across %d categories" % \
+      (signature, len(categories), len(category_count)))
+    self._tag("b", "#Items across categories: ")
+    self._text("%d (incl. possible duplicates)" % num_members)
+
+    self._br()
+    self._tag("b", "Span-level fact-matching statistics")
+    self.write_main_table_header(
+      ["Span Signature"],
+      [t.name for t in FactMatchType])
+    for span_signature, span_stats in span_match_counts.iteritems():
+      self._begin("tr")
+      self._cell(span_signature)
+      self._separator(header=False)
+      self.write_fact_match_counts(span_stats)
+      self._end("tr")
+    self._begin("tr")
+    self._cell("All")
+    self._separator(header=False)
+    self.write_fact_match_counts(match_counts)
+    self._end("tr")
+    self._end("table")
+
+    # Write the individual parses in a tabular form.
+    self._br()
+    self._tag("b", "Categories with parses matching '" + signature + "'")
+    seen = set()
+    max_rows = 200
```

Comment:
Probably fine with 200. Just wondering why 200? Might make sense to have it in a more central location if used elsewhere or simply in-lining it in the one place it is used. (Although the latter takes away the hint given by the name).

Reply:
It is unlikely we will want to navigate very deep in the list so I chose 200. But we can change that (or implement a pagination mechanism to see the next 200).